xtext/org.icyphy.linguafranca/src/org/icyphy/generator/GeneratorBase.xtend

```diff
@@ -408,49 +384,94 @@ abstract class GeneratorBase extends AbstractLinguaFrancaValidator {
-            this.resources.add(r.eResource)
-        }
-
-        for (r : this.resources) {
-            // Replace connections in this resources that are annotated with the 
-            // "after" keyword by ones that go through a delay reactor. 
-            r.insertGeneratedDelays(this)
-
-            if (r !== this.fileConfig.resource) {
-                // FIXME: create import graph to match import statements to erroneous resources.
-                // The following only works for direct imports...
-                val issues = (r as XtextResource).resourceServiceProvider.resourceValidator.validate(r, CheckMode.ALL,
-                    null)
-                for (issue : issues) {
-                    val imp = resource.allContents.filter(Import).
-                        findFirst[r.URI.toString().endsWith(it.importURI)] as Import
-                    reportError(imp, '''Unresolved compilation issues in '«imp.importURI»'.''')
-                }
-            }
-        }
-        // Assuming all AST transformations have completed, build the instantiation graph.
-        this.instantiationGraph = new InstantiationGraph(resource, false) // FIXME: obtain this from the validator instead
+        setResources(context)
+        
+        transformDelays()
 
-        // Topologically sort the reactors such that all of a reactor's
-        // dependencies occur earlier in the sorted list or reactors.
-        this.reactors = this.instantiationGraph.nodesInTopologicalOrder
+        // Invoke this function a second time because transformations may have introduced new reactors!
+        setReactorsAndInstantiationGraph()
 
         // First, produce any preamble code that the code generator needs
         // to produce before anything else goes into the code generated files.
         generatePreamble() // FIXME: Move this elsewhere.
     }
 
+
+    protected def setReactorsAndInstantiationGraph() {
+        // Assuming all AST transformations have completed, build the instantiation graph. Reactors introduced by
+        // AST transformations have to be included in this graph.
+        this.instantiationGraph = new InstantiationGraph(fileConfig.resource, false)
+
+        // Topologically sort the reactors such that all of a reactor's
+        // dependencies occur earlier in the sorted list or reactors.
+        this.reactors = this.instantiationGraph.nodesInTopologicalOrder
+        
+        // Add the main instance because it cannot be hooked into the model (syntax does not allow for it).
+        this.reactors.add(mainDef.reactorClass as Reactor)
+    }
+
+    /**
+     * For each involved resource, replace connections with delays with generated delay reactors.
+     */
+    protected def transformDelays() {
+         for (r : this.resources) {
+             r.insertGeneratedDelays(this)
+        }
+    }
+
+    /**
+     * Update the class variable that lists all the involved resources. Also report validation problems of imported 
```

Comment:
Perhaps add: A "resource" in Eclipse-speak is a file or a file within a jar file.

Reply:
OK